### PR TITLE
Refactored trilinos direct solver

### PR DESCRIPTION
**Pull request by mdh266**
Refactored the trilinos direct solver so that initialize() factorizes the matrix one an and overloaded solve() can be on the same linear system with multiple right handsides without factorizing the matrix every solve.  

Sorry about the two commits, my indentation was off on the first one.

### CODE REVIEW

**Comment by drwells** [comment]
It looks like there are still some whitespace issues.

Commits to fix the style are not a big deal :) If you would prefer, though, one way to get around it is to use `git rebase` to `fixup` your commits (that is, combine them and discard extra commit messages) and then do `git push -f` to overwrite the commits in the PR.

**Comment by mdh266** [comment]
Sure, would you like me to do that and resubmit it?

**Comment by Rombur** [comment]
If you ```rebase``` and then force the push this pull request will be updated. There is no need to resubmit.

**Comment by mdh266** [comment]
Okay.  I think I changed what you wanted. Sorry about that

**Comment by Rombur** [comment]
No need to apologies but it is still not right. You need to use the script ```./contrib/utilities/indent``` This will indent everything (you need to have astyle but it is very easy to install). Then, do a commit and use the command ```git rebase -i master```. This should open an editor where you can squash all your commits into one. Then, push on the branch. Git will complain that you rewrote the history (you went from multiple commits into one) so you will need to use the ```--force``` option.

File: include/deal.II/lac/trilinos_solver.h [hunk 1]
```diff
@@ -568,6 +568,24 @@ namespace TrilinosWrappers
     virtual ~SolverDirect ();
 
     /**
+     * Initializes the direct solver for the matrix <tt>A</tt> and creates a
+     * factorization for it with the package chosen from the additional 
+     * data structure. Note that there is no need for a preconditioner
+     * here and there is no solve preformred.
```
Comment (drwells) [inline include/deal.II/lac/trilinos_solver.h]: preformred -> performed

File: include/deal.II/lac/trilinos_solver.h [hunk 2]
```diff
@@ -568,6 +568,24 @@ namespace TrilinosWrappers
...
+     */
+     void
+     initialize (const SparseMatrix    &A);
+
+    /*
+     * Preforms the solve on the linear system <tt>Ax=b</tt> based of the 
+     * package set in itialize(). Note the matrix is not refactorized during
```
Comment (drwells) [inline include/deal.II/lac/trilinos_solver.h]: itialize -> initialize

File: source/lac/trilinos_solver.cc [hunk 3]
```diff
@@ -441,6 +441,86 @@ namespace TrilinosWrappers
...
+    // not.
+    ConditionalOStream  verbose_cout (std::cout,
+                                      additional_data.output_solver_details);
+
+    solver.reset();
+
+    // Next allocate the Amesos solver, this is done in two steps, first we
+    // create a solver Factory and and generate with that the concrete Amesos
+    // solver, if possible.
+    Amesos Factory;
```
Comment (drwells) [inline source/lac/trilinos_solver.cc]: The usual style convention is to make variables lower case.

Comment (mdh266) [inline source/lac/trilinos_solver.cc]: do you mean the A.trilinos?

**Comment by mdh266** [comment]
i dont have root access to install astyle.. but i tried using indent.. i hope that works

**Comment by bangerth** [comment]
It's still not quite right: https://travis-ci.org/dealii/dealii/builds/107340182

You can just put the `astyle` executable into a directory in your homedir and then add that directory to the `$PATH` environment variable. For example, on my machine, I've put it into `$HOME/bin` and then did
```
  export PATH=$PATH:$HOME/bin
```
When you then run `./contrib/utilities/indent`, it should work without error. You can see whether it worked by doing `git status` to see whether it modified any files. If it did, commit them and squash (or fix up) this last commit into the previous ones.

File: source/lac/trilinos_solver.cc (hunk 3 above)
Comment (drwells) [inline source/lac/trilinos_solver.cc]: Whoops, nevermind; that was there before. We can address that some other time.

**Comment by mdh266** [comment]
Hi, I'm getting  *** Found a version of astyle different than the required version 2.04.

and i used the setup_astyle.sh script found in contrib/utilities.. am I supposed to delete a different version or something?

**Comment by Rombur** [comment]
What probably happens is that another version of astyle is already installed on your system. Instead of using ```export PATH=$PATH:$HOME/bin``` like Wolfgang advised use ```export PATH=$HOME/bin:$PATH``` Now the system will first look in ```$HOME/bin``` and if it finds astyle there, it will use this version instead.

**Comment by mdh266** [comment]
Thanks that worked!

File: include/deal.II/lac/trilinos_solver.h [hunk 4]
```diff
@@ -33,11 +33,7 @@ DEAL_II_DISABLE_EXTRA_DIAGNOSTICS
 #  include <Epetra_Operator.h>
 #  include <Amesos.h>
 DEAL_II_ENABLE_EXTRA_DIAGNOSTICS
-
-
-DEAL_II_NAMESPACE_OPEN
-
-namespace TrilinosWrappers
+DEAL_II_NAMESPACE_OPEN namespace TrilinosWrappers
```
Comment (bangerth) [inline include/deal.II/lac/trilinos_solver.h]: Please undo this change.

File: include/deal.II/lac/trilinos_solver.h [hunk 5]
```diff
@@ -91,7 +88,7 @@ namespace TrilinosWrappers
        * base class for all solvers.
        */
       explicit
-      AdditionalData (const bool         output_solver_details   = false,
+      AdditionalData (const bool output_solver_details = false,
```
Comment (bangerth) [inline include/deal.II/lac/trilinos_solver.h]: What's the reason for this change?

File: include/deal.II/lac/trilinos_solver.h [hunk 6]
```diff
@@ -109,19 +106,18 @@ namespace TrilinosWrappers
...
-    SolverBase (SolverControl  &cn);
+    SolverBase (SolverControl &cn);
 
     /**
      * Second constructor. This constructor takes an enum object that
      * specifies the solver name and sets the appropriate Krylov method.
      */
-    SolverBase (const enum SolverName  solver_name,
-                SolverControl         &cn);
+    SolverBase (const enum SolverName solver_name, SolverControl &cn);
```
Comment (bangerth) [inline include/deal.II/lac/trilinos_solver.h]: same here

File: include/deal.II/lac/trilinos_solver.h [hunk 7]
```diff
@@ -109,19 +106,18 @@ namespace TrilinosWrappers
...
 
     /**
      * Destructor.
      */
-    virtual ~SolverBase ();
+    virtual ~ SolverBase ();
```
Comment (bangerth) [inline include/deal.II/lac/trilinos_solver.h]: ..and here, and multiple other places -- please go through the patch and make sure that you don't change things that don't need changing.

**Comment by bangerth** [comment]
@mdh266 -- the last two commits change a lot of things that don't need changing. Can you just throw both of these away (easily possible via `git rebase -i`) and run the indentation script only on the first few commits that actually changed stuff?

**Comment by mdh266** [comment]
Did this commit work out?  I think I properly indented it, but please let me know if I should do it again.

**Comment by davydden** [comment]
@mdh266  : it seems that you still have those unnecessary changes @bangerth pointed out above. What editor do you use? Perhaps it applies some automatic formatting which you should not use in this case. 
In particular, the last commit is a mixture of needed changes and those which are unnecessary. It would be good if you clean this up.

There is a way in Git how to work this out, but I assume you just started with Git and perhaps in this case it is better to do it by hand -- look at the original file and undo those unnecessary changes so that the difference between the two files is only the new feature you introduce.

File: include/deal.II/lac/trilinos_solver.h [hunk 8]
```diff
@@ -74,7 +70,8 @@ namespace TrilinosWrappers
      * one of the specialized derived classes when the solver should be set at
      * runtime. Currently enabled options are:
      */
-    enum SolverName {cg, cgs, gmres, bicgstab, tfqmr} solver_name;
```
Comment (davydden) [inline include/deal.II/lac/trilinos_solver.h]: Please undo this change.

File: include/deal.II/lac/trilinos_solver.h [hunk 9]
```diff
@@ -109,19 +106,18 @@ namespace TrilinosWrappers
     /**
      * Constructor. Takes the solver control object and creates the solver.
      */
-    SolverBase (SolverControl  &cn);
```
Comment (davydden) [inline include/deal.II/lac/trilinos_solver.h]: Please undo this change.

File: include/deal.II/lac/trilinos_solver.h [hunk 10]
```diff
@@ -129,10 +125,9 @@ namespace TrilinosWrappers
      * one of the linear solvers and preconditioners of Trilinos is chosen.
      */
     void
-    solve (const SparseMatrix     &A,
-           VectorBase             &x,
-           const VectorBase       &b,
-           const PreconditionBase &preconditioner);
```
Comment (davydden) [inline include/deal.II/lac/trilinos_solver.h]: Please undo this change.

File: include/deal.II/lac/trilinos_solver.h [hunk 11]
```diff
@@ -142,10 +137,9 @@ namespace TrilinosWrappers
      * Trilinos is chosen.
      */
     void
-    solve (Epetra_Operator        &A,
-           VectorBase             &x,
-           const VectorBase       &b,
-           const PreconditionBase &preconditioner);
```
Comment (davydden) [inline include/deal.II/lac/trilinos_solver.h]: Please undo this change.

File: include/deal.II/lac/trilinos_solver.h [hunk 12]
```diff
@@ -158,10 +152,10 @@ namespace TrilinosWrappers
      * will be thrown.
      */
     void
-    solve (const SparseMatrix           &A,
-           dealii::Vector<double>       &x,
-           const dealii::Vector<double> &b,
-           const PreconditionBase       &preconditioner);
```
Comment (davydden) [inline include/deal.II/lac/trilinos_solver.h]: Please undo this change.

File: include/deal.II/lac/trilinos_solver.h [hunk 13]
```diff
@@ -175,10 +169,10 @@ namespace TrilinosWrappers
      * exception will be thrown.
      */
     void
-    solve (Epetra_Operator              &A,
-           dealii::Vector<double>       &x,
-           const dealii::Vector<double> &b,
-           const PreconditionBase       &preconditioner);
```
Comment (davydden) [inline include/deal.II/lac/trilinos_solver.h]: Please undo this change.

File: include/deal.II/lac/trilinos_solver.h [hunk 14]
```diff
@@ -187,10 +181,10 @@ namespace TrilinosWrappers
      * solvers and preconditioners of Trilinos is chosen.
      */
     void
-    solve (const SparseMatrix                                  &A,
-           dealii::parallel::distributed::Vector<double>       &x,
-           const dealii::parallel::distributed::Vector<double> &b,
-           const PreconditionBase                              &preconditioner);
```
Comment (davydden) [inline include/deal.II/lac/trilinos_solver.h]: Please undo this change.

File: include/deal.II/lac/trilinos_solver.h [hunk 15]
```diff
@@ -200,23 +194,23 @@ namespace TrilinosWrappers
      * Trilinos is chosen.
      */
     void
-    solve (Epetra_Operator                                     &A,
-           dealii::parallel::distributed::Vector<double>       &x,
-           const dealii::parallel::distributed::Vector<double> &b,
-           const PreconditionBase                              &preconditioner);
```
Comment (davydden) [inline include/deal.II/lac/trilinos_solver.h]: Please undo this change.

**Comment by davydden** [comment]
@mdh266 : since you have so many unneeded changes, perhaps it is easier to: (i) copy `trilinos_solver.h` and `trilinos_solver.cc` (ii) overwrite them with original files (iii) re-introduce only the needed changes.

**Comment by bangerth** [comment]
I looked at this a while back and wanted to fix it myself but ran out of time that week.

The approach to do so is to just merge all of the changes made into one patch. (In fact, I think that the first commit by itself already contains everything that is necessary. The problem with the rest is that the commit history is not linear but is a diamond.) But there are other issues that need to address. First, the code is now replicated multiple times throughout the class. This should be cleaned up. Second, it would be good to have a test.

I did verify that the testsuite worked fine with this patch.

**Comment by davydden** [comment]
@bangerth you are right, the first commit https://github.com/dealii/dealii/pull/2141/commits/4ca9042ca832d5e31ea1024991328488cdcf3589 seems to have everything needed.  Even better, the pre-last one seems to be proper indented https://github.com/dealii/dealii/pull/2141/commits/5ab9dd41900c5eaafe552c8f11ec1c1c4c9daa8b and has the same info as the first one.

I think one should do on this branch  `git rebase -i HEAD~7` (there are seven commits in total here) and while editing the history remove all commits but the pre-last one. If one stash the current state all to one commit, there will be a lot of unneeded changes which are harder to cleanup. What do you say @bangerth ?

**Comment by mdh266** [comment]
Hi, I can definitely do whatever you need me to, just let me know what you decide is the best way to resolve the issue.

**Comment by luca-heltai** [comment]
Closed. Superseeded by #2493.
